### ISSUE TITLE
thanos compact cannot skip this error and I don't know how to fix it

### ISSUE THREAD

**Issue by gecailong** [improbable-eng/thanos#603]
thanos version:
version 0.1.0
prometheus version:
version 2.3.1

problem：
Thanos compact has been running. One day, the following error occurred. I don't know if there is any problem with the data in ceph，I tried to delete the file, but it didn't work。Now thanos compact cannot be started because of this error
**Error  log:**
level=error ts=2018-10-31T08:46:54.179754425Z caller=compact.go:212 msg="retriable error" err="compaction failed: sync: retrieve bucket block metas: downloading meta.json for 01CT26WKHAENAMN9XAFM0HS642: meta.json bkt get for 01CT26WKHAENAMN9XAFM0HS642: The specified key does not exist."

**Resultion**:
Since the data is about 100T per day, I urgently need compact
and Any good solutions?

**Comment by bwplotka** [comment]
Can you provide more info, how your setup is configured? All those block metas are required to tell what's is required to be compacted.

> I tried to delete the file, but it didn't work。

Which file? You mean whole block `01CT26WKHAENAMN9XAFM0HS642`?

**Comment by antonio** [comment]
:wave:

I think I'm also seeing this, only not only in the compactor, but also in the store nodes. From what I think I've understood after analyzing the logs for all the components and the chain of events seems to be:

1. The compactor created a new block by compacting 6 existing blocks

```
level=info ts=2018-11-15T03:24:34.972666428Z caller=compact.go:347 msg="compact blocks" count=6 mint=1539475200000 maxt=1539648000000 ulid=01CWAPC3BZVVXXDHWBXYG3V82F sources="[01CT71T5SZ69J41PJYH5CJXDQT 01CT8CCBRWF3PAYN366FN0E7GH 01CT9KZCYMS26A9M868Q5FX5MX 01CTAYS9NC61NKQKP6T5EG4XW8 01CTC6CSD1MAFW89CPA05758FH 01CTDH74R15V09NA53XSVH2MCM]"
```

2. The compactor deleted the original blocks

```
level=info ts=2018-11-15T03:24:45.099017858Z caller=compact.go:775 compactionGroup="0@{<REDACTED>}" msg="deleting compacted block" old_block=01CT71T5SZ69J41PJYH5CJXDQT result_block=01CWAPC3BZVVXXDHWBXYG3V82F
```

3. From that time onward, all the stores (theres's 3, load balanced with consul) and the compactor return errors regarding the deleted block. The error message is exactly the same in both components.

```
level=warn ts=2018-11-15T03:36:15.264028035Z caller=bucket.go:240 msg="loading block failed" id=01CT71T5SZ69J41PJYH5CJXDQT err="new bucket block: load meta: download meta.json: get file: The specified key does not exist."
```

For what is worth, the block does no longer exist in the bucket (we use S3) nor the data folder for the stores/compactor.

**Comment by bwplotka** [comment]
@antonio it looks like this for store: https://github.com/improbable-eng/thanos/issues/564 

But it's weird that compactor shows same error? Worth to investigate in code.

But all of this should not block compactor/store from doing its work.

**Comment by gecailong** [comment]
> But it's weird that compactor shows same error? Worth to investigate in code.
> 
> But all of this should not block compactor/store from doing its work.

yes,you are right! I also don't think he should affect the running of the program

**Comment by antonio** [comment]
> But all of this should not block compactor/store from doing its work.

It did in our case because the compactor exited right after that, and didn't compact anything most of the time. Since then we are using the `-w` flag to make sure that an error does not abort the compaction and things look much better.
